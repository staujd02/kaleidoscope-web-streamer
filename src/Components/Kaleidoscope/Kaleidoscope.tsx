import React from 'react';
import Loader from '../Loader/Loader';
import Cycler from '../Cycler/Cycler';
import Configuration from '../Configuration/Configuration';
import SourceRepo from '../../LocalStorage/SourceRepo';

export default class Kaleidoscope extends React.Component<KaleidoscopeProps, KaleidoscopeState>{

    constructor(props: KaleidoscopeProps) {
        super(props);
        this.state = {
            isConfiguring: false,
            sourceRepository: {
                streams: []
            }
        }
    }

    componentDidMount() {
        this.setState({
            sourceRepository: SourceRepo.load()
        });
    }

    openConfiguration = (event: ButtonClickEvent): void => {
        this.setState({
            isConfiguring: true
        });
    }

    handleSave = (updatedSourceRepo: SourceRepository) => {
        SourceRepo.save(updatedSourceRepo);
        this.setState({
            sourceRepository: updatedSourceRepo
        });
    }

    doneConfiguringCallback = () => {
        this.setState({
            isConfiguring: false
        });
    }

    render() {
        const { isConfiguring, sourceRepository } = this.state;
        let sources = sourceRepository.streams.filter(s => s.isEnabled);
        sources.sort((a, b) => a.sortOrder - b.sortOrder);
        return (
            <React.Fragment>
                {isConfiguring && <Configuration doneConfiguring={this.doneConfiguringCallback} sourceRepo={sourceRepository} handleSave={this.handleSave} />}
                {!isConfiguring &&
                    <Loader loadTime={this.props.loadTime} handleConfigureClick={this.openConfiguration}>
                        <Cycler openMenuCallback={this.openConfiguration} sourceList={sources} />
                    </Loader>
                }
            </React.Fragment>
        );
    }
}