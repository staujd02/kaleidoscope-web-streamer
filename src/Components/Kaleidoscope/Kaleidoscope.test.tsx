import React from 'react';
import { ShallowWrapper, shallow } from 'enzyme';
import Kaleidoscope from './Kaleidoscope';
import { formatHTML } from '../../TestUtilities/htmlFormatter';

describe('The Kaleidoscope', () => {

    let wrapper: ShallowWrapper<KaleidoscopeProps, KaleidoscopeState, Kaleidoscope>

    beforeEach(() => {
        establishMock();
        wrapper = shallow(<Kaleidoscope loadTime={1} />);
    });

    it('renders correctly', () => expect(formatHTML(wrapper.html())).toMatchSnapshot());

    describe('when the component mounts', () => {
        describe('given the source repo is empty', () => {
            it('should load the streams from local storage', () => {
                expect(wrapper.state().sourceRepository).toEqual({
                    streams: [{
                        title: 'Title 1',
                        sortOrder: 1,
                        source: "http1",
                        isEnabled: true,
                        duration: 50
                    }]
                });
            });
        });
    });

    describe('given the user is configuring the streams', () => {
        beforeEach(() => {
            wrapper.setState({
                isConfiguring: true
            });
        });

        it('renders correctly',
            () => expect(formatHTML(wrapper.html())).toMatchSnapshot());
    });

    function establishMock() {
        jest.mock('../../LocalStorage/SourceRepo', () => {
            return {
                load: jest.fn().mockReturnValue({
                    streams: [
                        {
                            title: 'Title 1',
                            sortOrder: 1,
                            source: "http1",
                            isEnabled: true,
                            duration: 50
                        }
                    ]
                } as SourceRepository),
                save: jest.fn()
            }
        });
    }

    afterAll(() => {
        jest.unmock('../../LocalStorage/SourceRepo');
    });

});