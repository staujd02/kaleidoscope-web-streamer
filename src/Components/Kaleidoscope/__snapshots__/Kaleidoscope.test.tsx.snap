// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`The Kaleidoscope given the user is configuring the streams renders correctly 1`] = `
"
<header class=\\"config-viewport\\">
  <div class=\\"config-container\\">
    <button class=\\"MuiButtonBase-root MuiButton-root cycle MuiButton-text\\" tabindex=\\"0\\" type=\\"button\\">
      <span class=\\"MuiButton-label\\">Resume Cycle</span>
    </button>
    <div class=\\"MuiPaper-root MuiPaper-elevation1 MuiCard-root card MuiPaper-rounded\\">
      <div class=\\"MuiCardContent-root\\">
        <p class=\\"MuiTypography-root title MuiTypography-body1 MuiTypography-colorTextSecondary MuiTypography-gutterBottom\\">List of the Streams</p>
        <div class=\\"MuiFormControl-root MuiTextField-root search MuiFormControl-marginDense\\">
          <label class=\\"MuiFormLabel-root MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-marginDense MuiInputLabel-outlined\\" data-shrink=\\"false\\" for=\\"outlined-dense\\">Search</label>
          <div class=\\"MuiInputBase-root MuiOutlinedInput-root MuiInputBase-formControl MuiInputBase-marginDense MuiOutlinedInput-marginDense\\">
            <input type=\\"text\\" aria-invalid=\\"false\\" class=\\"MuiInputBase-input MuiOutlinedInput-input MuiInputBase-inputMarginDense MuiOutlinedInput-inputMarginDense\\" id=\\"outlined-dense\\">
            <fieldset aria-hidden=\\"true\\" style=\\"padding-left:8px\\" class=\\"PrivateNotchedOutline-root-137 MuiOutlinedInput-notchedOutline\\">
              <legend class=\\"PrivateNotchedOutline-legend-138\\" style=\\"width:0.01px\\">
                <span>â€‹</span>
              </legend>
            </fieldset>
          </div>
        </div>
        <ul class=\\"MuiList-root MuiList-padding\\" aria-label=\\"streams\\">
          <div class=\\"MuiButtonBase-root MuiListItem-root MuiListItem-gutters MuiListItem-button\\" tabindex=\\"0\\" role=\\"button\\" aria-disabled=\\"false\\">
            <div class=\\"MuiListItemText-root\\">
              <span class=\\"MuiTypography-root MuiListItemText-primary MuiTypography-body1\\">Title 1</span>
            </div>
          </div>
        </ul>
      </div>
    </div>
    <div class=\\"MuiPaper-root MuiPaper-elevation1 MuiCard-root card MuiPaper-rounded\\">
      <div class=\\"MuiCardContent-root\\">
        <p class=\\"MuiTypography-root title MuiTypography-body1 MuiTypography-colorTextSecondary MuiTypography-gutterBottom\\">Stream Details</p>
        <p class=\\"MuiTypography-root placeholder MuiTypography-body1\\">Select a stream to view it's details.</p>
      </div>
    </div>
    <div class=\\"MuiPaper-root MuiPaper-elevation1 MuiCard-root card MuiPaper-rounded\\">
      <div class=\\"MuiCardContent-root\\">
        <p class=\\"MuiTypography-root title MuiTypography-body1 MuiTypography-colorTextSecondary MuiTypography-gutterBottom\\">Preview of the Stream</p>
        <p class=\\"MuiTypography-root placeholder MuiTypography-body1\\">Select a stream to view a preview of it.</p>
      </div>
    </div>
  </div>
</header>
"
`;

exports[`The Kaleidoscope renders correctly 1`] = `
"
<header class=\\"App-header\\">
  <img src=\\"kaleidoscope-2.svg\\" class=\\"App-logo\\" alt=\\"logo\\">
  <p>Starting Up The Streams...</p>
  <button class=\\"MuiButtonBase-root MuiButton-root configure-button MuiButton-text MuiButton-textPrimary\\" tabindex=\\"0\\" type=\\"button\\">
    <span class=\\"MuiButton-label\\">Configure Streams</span>
  </button>
</header>
"
`;
